src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/SemanticTokensCache.cs

```diff
@@ -0,0 +1,216 @@
+
+        /// <summary>
+        /// Caches tokens on a per-line basis. If the given line has already been cached for the document
+        /// and its semantic version, the line's tokens will be ignored.
+        /// </summary>
+        /// <param name="uri">The URI associated with the passed-in tokens.</param>
+        /// <param name="semanticVersion">The document's semantic version associated with the passed-in tokens.</param>
+        /// <param name="range">The range associated with the passed-in tokens.</param>
+        /// <param name="tokens">The complete set of tokens for the passed-in range.</param>
+        internal void CacheTokens(
+            DocumentUri uri,
+            VersionStamp semanticVersion,
+            Range range,
+            int[] tokens)
+        {
+            if (!_cache.TryGetValue(uri, out var documentCache))
+            {
+                documentCache = new MemoryCache<VersionStamp, Dictionary<int, IReadOnlyList<int>>>(sizeLimit: MaxSemanticVersionPerDoc);
+                _cache.Set(uri, documentCache);
+            }
+
+            if (!documentCache.TryGetValue(semanticVersion, out var lineToTokensDict))
+            {
+                lineToTokensDict = new Dictionary<int, IReadOnlyList<int>>();
+                documentCache.Set(semanticVersion, lineToTokensDict);
+            }
+
+            var absoluteLine = 0;
+
+            // Cache the tokens associated with each line
+            for (var tokenIndex = 0; tokenIndex < tokens.Length; tokenIndex += 5)
+            {
+                absoluteLine += tokens[tokenIndex];
+
+                // Don't cache tokens outside of the given range
+                if (absoluteLine < range.Start.Line)
+                {
+                    continue;
+                }
+
+                if (absoluteLine >= range.End.Line)
+                {
+                    break;
+                }
+
+                if (lineToTokensDict.TryGetValue(absoluteLine, out _))
+                {
+                    // This line is already cached, so we can skip it
+                    continue;
+                }
+
+                var lineTokens = new List<int>();
+
+                // Cache the line's tokens
+                while (tokenIndex < tokens.Length)
+                {
+                    lineTokens.Add(tokens[tokenIndex]);
+                    lineTokens.Add(tokens[tokenIndex + 1]);
+                    lineTokens.Add(tokens[tokenIndex + 2]);
+                    lineTokens.Add(tokens[tokenIndex + 3]);
+                    lineTokens.Add(tokens[tokenIndex + 4]);
+
+                    if (tokenIndex + 5 >= tokens.Length || tokens[tokenIndex + 5] != 0)
+                    {
+                        // We've hit the next line or the end of the tokens
+                        break;
+                    }
+
+                    tokenIndex += 5;
+                }
```

Comment:
```suggestion
                // Loop until we hit the next line or the end of the tokens
                do
                {
                    lineTokens.Add(tokens[tokenIndex]);
                    lineTokens.Add(tokens[tokenIndex + 1]);
                    lineTokens.Add(tokens[tokenIndex + 2]);
                    lineTokens.Add(tokens[tokenIndex + 3]);
                    lineTokens.Add(tokens[tokenIndex + 4]);
                    
                    tokenIndex += 5;
                }
                while (tokenIndex < tokens.Length && tokens[tokensIndex] == 0);
```

Reply:
alternatively:
```suggestion
                // Loop until we hit the next line or the end of the tokens
                do
                {
                    lineTokens.Add(tokens[tokenIndex++]);
                    lineTokens.Add(tokens[tokenIndex++]);
                    lineTokens.Add(tokens[tokenIndex++]);
                    lineTokens.Add(tokens[tokenIndex++]);
                    lineTokens.Add(tokens[tokenIndex++]);
                }
                while (tokenIndex < tokens.Length && tokens[tokensIndex] == 0);
```

Reply:
Hmm, I don't know if the suggested options would be equivalent. The original version only increments `tokenIndex` if the next token is on the same line, but the two suggested versions increment `tokenIndex` regardless if the next token is on the same line or not.

Reply:
right.  but this is because you have an outer `tokenIndex += 5` in your for loop which feels suspect to me.  let's chat offline, i have some suggestions on how to make this read more clearly.

Reply:
In my latest push I made the outer loop a while loop and took your first suggested change. Hopefully it's more readable now?